### PR TITLE
Migrate sending metrics to Cloud-platform-wide Google Analytics

### LINKED ISSUE #693: Additionally send usage data to the unified internal metrics collection system
We currently send our usage data to a GA property exclusive to IntelliJ. Ultimately, we want to migrate this traffic to the centralized internal metrics collection system, where analysts can query various Cloud-related metrics across all products, sites, marketing campaigns, documents, etc.

Until almost all of our plugin users complete upgrade, we need to send usage data to both the current IntelliJ-specific Google Analytics (GA) property and the centralized GA property during the transition time. Eventually, we will turn down the IntelliJ-specific GA property.

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #693.

I tried to re-use the existing facilities for sending usage data as much as possible. In the long run, we would probably want to overhaul this.

I will be doing actual ping tests in various ways.

### CODE REVIEW

**Comment by chanseokoh** [comment]
@loosebazooka Can you also look at `build.gradle`? I did not really understand why `input.property()` was used previously; it was just passing an independent map to `expand()`.

**Comment by patflynn** [comment]
why not just switch entirely?

**Comment by chanseokoh** [comment]
@patflynn Actually, I would love to stop sending usage information to the old (i.e., current) GA right away. However, if we decided to send pings to the new GA property only, we will see partial traffic coming only from users who updated our plugin. If this is ok, I can really gut out the old code and start fresh. (Perhaps we could aggregate the old GA and the new GA traffic at the dashboard level though. Not tried this yet.)

**Comment by patflynn** [comment]
Yeah let's not keep the old code path. Updated plugins will go to the new
project, and non-updated plugins will go to our old project. That's fine.

**Comment by chanseokoh** [comment]
Question: if on Android Studio, will we send pings only to the Android GA property, or to both Android GA and Cloud GA?

**Comment by chanseokoh** [comment]
To clarify my last question, my understanding is that the Android Studio team has their own GA property (which I am not aware of and have no access to). @loosebazooka, is this correct? And, probably, every metrics event happening from our code on Android Studio goes to that GA property, doesn't it? So, the answer to my last question is probably 'No, and keep sending pings to Android GA only', right?

For IntelliJ, we also have our own GA property currently, but we will abandon it and migrate to use the unified Cloud GA property.

**Comment by chanseokoh** [comment]
@patflynn @etanshaul PTAL.

It turns out that we cannot just get rid of the old code since our code may be synced with Android Studio later. We decided not to change the `trackEvent()` interface at the moment. Considering all this, I think it is best in many aspects to just modify existing code a bit to achieve our goal.

To recap, we will now send pings to the Cloud-platform-wide GA Property. Consequently, users who don't upgrade will continue to send pings to the old (i.e., current) GA Property, which we will eventually deprecate.

**Comment by etanshaul** [comment]
Is there some background info on this you could point me to if you'd like me to review this? I don't have the info on what needs to be collected and where it needs to be sent.

**Comment by chanseokoh** [comment]
I think for efficiency and productivity, it's best for you to review just the syntactic structure and programmatic soundness here. I'll ensure the correctness of the functionality of sending metrics to the Cloud GA. I tested and verified that I can see my pings appearing in the target GA Property.

However, if you really want to go deep in reviewing, I think you can take a look at [this implementation](https://github.com/google/cloud-reporting/blob/master/src/main/java/com/google/cloud/metrics/MetricsUtils.java#L183). I basically referenced that implementation along with [the GA protocol and parameter reference].(https://developers.google.com/analytics/devguides/collection/protocol/v1/reference).

**Comment by chanseokoh** [comment]
And feel free to raise any suspicion or ask any question if are curious about anything, even if it's not related to the review task.

File: google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java [hunk 1]
```diff
@@ -36,64 +34,76 @@
...
+   * @param eventAction On IntelliJ, this is the primary event action.
+   * @param eventLabel On IntelliJ, reported as a metadata key.
+   * @param eventValue On IntelliJ, reported as a metadata value.
+   */
   @Override
   public void trackEvent(@NotNull String eventCategory,
       @NotNull String eventAction,
       @Nullable String eventLabel,
       @Nullable Integer eventValue) {
+    if (eventValue != null) {
```
Comment (etanshaul) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: whats the point of this if statement if you are using preconditions#checkArgument right after?

Comment (etanshaul) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: maybe the better question is what was your intent here? the if statement wrapping the precondition check is confusing.

Comment (chanseokoh) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: I could then do `checkArgument(eventValue == null || eventLabel != null)`. That is, if `eventValue` is defined, it must be that `eventLabel` is defined as well. Will this work?

Comment (etanshaul) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: Oh I misread this, I didn't catch that the precondition is checking a different variable. Maybe instead do it like this for readability (if it has the same logic):

`Preconditions.checkArgument(eventValue != null && eventLabel != null)`

Comment (etanshaul) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: Thats actually even harder to read to me. Just leave it how you had it now that I understand what you meant. Just to confirm: if eventValue is defined but eventLabel is null, do you really want to throw an exception?

Comment (chanseokoh) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: I can decide with some discretion, so I will enforce this rule that eventLabel-eventValue is basically a key-value metadata pair. In this sense, a value without key doesn't make much sense.

Comment (etanshaul) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: agree that it doesn't make sense, but it feels arbitrary that you are doing a precondition on the label (and not the value for example). why not, when you are consuming them, just guard against `eventValue != null && eventLabel != null`. This isn't that big of a deal, but it just made the code confusing (to me at least).

Comment (chanseokoh) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: Sounds good. I will just guard the invalid case with a if-statement.

Comment (etanshaul) [inline google-account-plugin/src/com/google/cloud/tools/intellij/stats/GoogleUsageTracker.java]: Other than that, lgtm without having looked into the analytics details.

**Comment by patflynn** [comment]
Just curious if you've tested this out with analytics opted in?

**Comment by patflynn** [comment]
LGTM otherwise

**Comment by chanseokoh** [comment]
Yes, I set a test (sort of) GA tracking ID (with the $ANALYTICS_ID env) on my machine and opted in on my local run/debug sessions. I can see pings appear on the target GA.

**Pull request merged**